packages/vm/src/imports.rs

```diff
@@ -150,19 +147,29 @@ macro_rules! write_region {
 }
 
 /// Reads a storage entry from the VM's storage into Wasm memory
-pub fn do_read<S: Storage, Q: Querier>(
-    ctx: &mut Ctx,
-    key_ptr: u32,
-    value_ptr: u32,
-) -> VmResult<i32> {
-    let key = read_region!(ctx, key_ptr, MAX_LENGTH_DB_KEY);
+pub fn do_read<S: Storage, Q: Querier>(ctx: &mut Ctx, key_ptr: u32) -> VmResult<u32> {
+    let key = read_region(ctx, key_ptr, MAX_LENGTH_DB_KEY)?;
     // `Ok(expr?)` used to convert the error variant.
     let value: Option<Vec<u8>> =
         with_storage_from_context::<S, Q, _, _>(ctx, |store| Ok(store.get(&key)?))?;
-    Ok(match value {
-        Some(buf) => write_region!(ctx, value_ptr, &buf),
-        None => errors::read::KEY_DOES_NOT_EXIST,
-    })
+
+    let out_data = match value {
+        Some(data) => data,
+        None => return Ok(0),
+    };
+
+    let out_ptr = with_func_from_context::<S, Q, u32, u32, _, _>(ctx, "allocate", |allocate| {
+        let out_size = to_u32(out_data.len())?;
+        let ptr = allocate.call(out_size)?;
+        if ptr == 0 {
+            return Err(make_generic_err(
+                "Got a null Wasm pointer from allocate(). This is not allowed by CosmWasm even if it is valid Wasm. The contract needs to ensure to allocate at a non-null address. See https://stackoverflow.com/q/62042733/2013738 for more information.",
+            ));
+        }
```

Comment:
I would move this check to outside the closure, just to keep things more readable.